libgalois/src/analytics/bfs/bfs.cpp

```diff
@@ -147,68 +103,60 @@ AsynchronousAlgo(
     katana::LargeArray<Dist>* node_data, const P& pushWrap,
     const R& edgeRange) {
   namespace gwl = katana;
+
+  using Loop = typename std::conditional<
+      CONCURRENT, katana::ForEach, katana::StdForEach>::type;
   using FIFO = gwl::PerSocketChunkFIFO<kChunkSize>;
   using BSWL = gwl::BulkSynchronous<gwl::PerSocketChunkLIFO<kChunkSize>>;
   using WL = FIFO;
 
-  using Loop = typename std::conditional<
-      CONCURRENT, katana::ForEach, katana::WhileQ<katana::SerFIFO<T>>>::type;
-
   KATANA_GCC7_IGNORE_UNUSED_BUT_SET
   constexpr bool use_CAS = CONCURRENT && !std::is_same<WL, BSWL>::value;
   KATANA_END_GCC7_IGNORE_UNUSED_BUT_SET
 
-  Loop loop;
-
+  // TODO(lhc): How about `failed_work`?
   katana::GAccumulator<size_t> bad_work;
   katana::GAccumulator<size_t> WL_empty_work;
 
-  (*node_data)[source] = 0;
+  (*node_data)[source] = source;
   katana::InsertBag<T> init_bag;
 
   if (CONCURRENT) {
-    pushWrap(init_bag, source, 1, "parallel");
+    pushWrap(init_bag, source, "parallel");
   } else {
-    pushWrap(init_bag, source, 1);
+    pushWrap(init_bag, source);
   }
 
+  Loop loop;
+
   loop(
       katana::iterate(init_bag),
       [&](const T& item, auto& ctx) {
-        const Dist& sdist = (*node_data)[item.src];
-
-        if (kTrackWork) {
-          if (item.dist != sdist) {
-            WL_empty_work += 1;
-            return;
-          }
-        }
-
-        const Dist new_dist = item.dist;
+        const GNode parent_candidate = item.src;
 
         for (auto ii : edgeRange(item)) {
           auto dest = graph.GetEdgeDest(ii);
-          Dist& ddata = (*node_data)[*dest];
+          GNode& parent = (*node_data)[*dest];
 
           while (true) {
-            Dist old_dist = ddata;
-            if (old_dist <= new_dist) {
+            GNode current_parent = parent;
+            if (current_parent != BfsImplementation::kDistanceInfinity) {
```

Comment:
@nicelhc13 : This is asynchronous algorithm and without checking on lower distance, it's wrong. If we're going level by level, it's fine, otherwise first CAS wins, which could be from a node with higher distance.

Reply:
I see. That makes sense to me. I thought that operator is visiting regardless of distance, in this case. If it is the case, do you have any suggestion for asynchronous running? Should I add the second label called parent and also fill this?

Reply:
Let's discuss this via slack first.